### ISSUE TITLE
azure service principal is not supported to authorize in kpack0.6.0 and kpack0.6.1

### ISSUE THREAD

**Issue by smile37773** [pivotal/kpack#1005]
The clusterstore failed due to unauthorized. It seems the service principal is not supported to authorize and only MSI is supported.

Here is the error detail of clusterstore:
message: 'GET https://serviceacr.azurecr.io/oauth2/token?scope=repository%3Apaketo-buildpacks%2Fjava-native-image%3Apull&service=serviceacr.azurecr.io:
UNAUTHORIZED: authentication required, visit https://aka.ms/acr/authorization
for more information.'
status: "False"
type: Ready
observedGeneration: 1

We already set the AZURE_CONTAINER_REGISTRY_CONFIG in kpack-controller.

**Comment by tylerphelan** [comment]
This should be supported. Use the service principal's id and password as the default repository credential username and password as described [here](https://github.com/pivotal/kpack/blob/main/docs/tutorial_kp.md).

**Comment by smile37773** [comment]
@tylerphelan  They are two different credentials. The service principal used in clusterstack is configured in /etc/kubernetes/azure.json.  We set the AZURE_CONTAINER_REGISTRY_CONFIG in kpack-controller and mount the file.

like this:
containers:
name: controller
image: {{controller-image}}
env:
name: ENABLE_PRIORITY_CLASSES
value: "false"
**name: AZURE_CONTAINER_REGISTRY_CONFIG
value: /etc/kubernetes/azure.json**
name: CONFIG_LOGGING_NAME
value: config-logging
name: CONFIG_OBSERVABILITY_NAME
value: config-observability
name: METRICS_DOMAIN
value: kpack.io
name: SYSTEM_NAMESPACE
valueFrom:
fieldRef:
fieldPath: metadata.namespace

**Comment by tylerphelan** [comment]
@smile37773 
this was working previously before kpack 0.6.0?

**Comment by smile37773** [comment]
@tylerphelan  Yes. It can work before but not in kpack 0.6.0 and kpack 0.6.1.

**Comment by tylerphelan** [comment]
@smile37773 would y'all be able to try using the environment variables listed [here](https://github.com/chrismellard/docker-credential-acr-env/) instead of the previous `AZURE_CONTAINER_REGISTRY_CONFIG`?

```
AZURE_CLIENT_ID=<clientID>
AZURE_CLIENT_SECRET=<clientSecret>
AZURE_TENANT_ID=<tenantId>
```

**Comment by smile37773** [comment]
Thanks @tylerphelan  using the environment variables can work. But could you please find out why it doesn't work with `AZURE_CONTAINER_REGISTRY_CONFIG`. We need this environment because we not only use the service principal but also use the MSI. The `AZURE_CONTAINER_REGISTRY_CONFIG` is necessary for us.

**Comment by tylerphelan** [comment]
@smile37773 this change is due to the underlying registry client library https://github.com/google/go-containerregistry changed its ACR helper dependency from a deprecated https://github.com/vdemeester/k8s-pkg-credentialprovider/ which uses the azure.json file to https://github.com/chrismellard/docker-credential-acr-env/ which uses the above env vars. The behavior should be the same if you use the env vars instead of a file.

**Comment by smile37773** [comment]
@tylerphelan  I see it still import https://github.com/vdemeester/k8s-pkg-credentialprovider/ [here](https://github.com/pivotal/kpack/blob/ed7fe5056666f0ec1aa4378c92db1d52a0a872c5/pkg/dockercreds/k8sdockercreds/azurecredentialhelperfix/setup.go)

**Comment by tylerphelan** [comment]
@smile37773 thank you! we missed removing this dependency in our last release and will remove it in the next release of kpack ([PR](https://github.com/pivotal/kpack/pull/1010)). However, it is effectively unused and will not affect behavior.

**Comment by zhoufenqin** [comment]
@tylerphelan what should we do after removing `k8s-pkg-credentialprovider`, should we config `AZURE_CLIENT_ID/AZURE_CLIENT_SECRET/AZURE_TENANT_ID` in kpack-controller deployment manually? I think it's not a good idea to config these settings to kpack-controller deployment, we should use ` /etc/kubernetes/azure.json` in VM node to determine if the AKS cluster is using `ServicePrinciple` or `MSI`
https://docs.microsoft.com/en-us/azure/aks/kubernetes-service-principal?tabs=azure-cli

**Comment by smile37773** [comment]
@tylerphelan  we still need this env `AZURE_CONTAINER_REGISTRY_CONFIG`.  Could you please read the secret and id from azure.json and then set the env `AZURE_CLIENT_ID` in kpack-controller. It doesn't make sense the `AZURE_CONTAINER_REGISTRY_CONFIG` has been deprecated and it has blocked us to upgrade kpack version.

**Comment by tylerphelan** [comment]
@smile37773 what is the benefit of using the `azure.json` file over the environment variables? 

Using the environment variables appears to be a somewhat accepted convention and is used in cloud native projects such as [ko](https://github.com/google/ko/blob/691da5afb64e1d9204b250abc41fc738737b1ebb/pkg/commands/config.go#L47) and [cosign](https://github.com/sigstore/cosign/blob/516cc39f5582f46af9fc5b9081af08dba24441e7/cmd/cosign/cli/options/registry.go#L82-L90)

**Comment by smile37773** [comment]
@tylerphelan  The benefit is that we don't need to do the migration. The `azure.json` has two configurations which are service principal and MSI. We don't care which configuration is it. We only set the env lin kpack-controller like this:
name: AZURE_CONTAINER_REGISTRY_CONFIG
value: /etc/kubernetes/azure.json
However, if we can't use `AZURE_CONTAINER_REGISTRY_CONFIG`, we should provide two kinds of kpack-controller. One for service principal, the other for MSI.

**Comment by showpune** [comment]
@tylerphelan there are some problem to using the env instead of azure.json
1) Customer has different auth model ( service principal and MSI ),  we need to set different env during deployment in different auth mode
2) There should be secret rotation issue if using env
3) We need to migrate all existed service

It is better to make the auth transparent to App

**Comment by tylerphelan** [comment]
@showpune @smile37773 we released a patch to fix this regression in kpack 0.6.2. Could you please give it a try and let us know if it works for y'all!

https://github.com/pivotal/kpack/releases/tag/v0.6.2

**Comment by smile37773** [comment]
@tylerphelan Thank you, it works.
